test/performance/odp_crc.c

```diff
@@ -0,0 +1,238 @@
+/* Copyright (c) 2021, Nokia
+ *
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <getopt.h>
+
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+
+/* Command line options */
+typedef struct {
+	uint32_t time;
+} options_t;
+
+static options_t options;
+static const options_t options_def = {
+	.time = 1000,
+};
+
+static void print_usage(void)
+{
+	printf("\n"
+	       "CRC performance test\n"
+	       "\n"
+	       "Usage: odp_crc_perf [options]\n"
+	       "\n"
+	       "  -t, --time    Approximate time (ms) for each measurement (default %u)\n"
+	       "  -h, --help    This help\n"
+	       "\n",
+	       options_def.time);
+}
+
+static int parse_options(int argc, char *argv[])
+{
+	int opt;
+	int long_index;
+	int ret = 0;
+
+	static const struct option longopts[] = {
+		{ "time", required_argument, NULL, 't' },
+		{ "help", no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 }
+	};
+
+	static const char *shortopts = "+t:h";
+
+	options = options_def;
+
+	while (1) {
+		opt = getopt_long(argc, argv, shortopts, longopts, &long_index);
+
+		if (opt == -1)
+			break;
+
+		switch (opt) {
+		case 't':
+			options.time = atol(optarg);
+			break;
+		case 'h':
+			/* fall through */
+		default:
+			print_usage();
+			ret = -1;
+			break;
+		}
+	}
+
+	if (options.time < 1) {
+		ODPH_ERR("Invalid time: %" PRIu32 "\n", options.time);
+		return -1;
+	}
+
+	return ret;
+}
+
+static uint8_t *buf, *data;
+static const int bufsize = 32 * 1024;
+
+static void prepare_data(void)
+{
+	uint64_t seed = 1;
+	const unsigned long page = 4096;
+
+	/* One extra page for alignment, another one for offset. */
+	buf = (uint8_t *)malloc(bufsize + page * 2);
+	data = (uint8_t *)(((uintptr_t)buf + (page - 1)) & ~(page - 1));
```

Comment:
ODPH_ERR() and exit if malloc() fails.

Usage of SHM would be better than malloc, if bufsize would be increased later on (SHM uses likely hugepages for large allocs).

Reply:
> ODPH_ERR() and exit if malloc() fails.

Done in v2.

I vote that we switch to shm later if needed. Either way, above 32K buffer size this increasingly becomes a memory read speed test instead of crc speed test.

Reply:
It's OK to use malloc for now.